hazelcast/include/hazelcast/client/serialization/pimpl/compact.h

```diff
@@ -0,0 +1,426 @@
+{};
+
+/**
+ * Provides means of reading compact serialized fields from the binary data.
+ * <p>
+ * Read operations might throw hazelcast_serialization exception.
+ * when a field with the given name is not found or there is a type mismatch. On
+ * such occasions, one might provide default values to the read methods to
+ * return it in case of the failure scenarios described above. Providing default
+ * values might be especially useful, if the class might evolve in future,
+ * either by adding or removing fields.
+ *
+ * @since 5.2
+ */
+class HAZELCAST_API compact_reader
+{
+public:
+    /**
+     * Reads a 32-bit two's complement signed integer.
+     *
+     * @param fieldName name of the field.
+     * @return the value of the field.
+     * @throws hazelcast_serialization if the field does not exist in
+     * the schema or the type of the field does not match with the one defined
+     * in the schema.
+     */
+    int32_t read_int32(const std::string& field_name);
+
+    /**
+     * Reads a 32-bit two's complement signed integer or returns the default
+     * value.
+     *
+     * @param fieldName    name of the field.
+     * @param defaultValue default value to return if the field with the given
+     * name does not exist in the schema or the type of the field does not match
+     * with the one defined in the schema.
+     * @return the value or the default value of the field.
+     */
+    int32_t read_int32(const std::string& field_name, int32_t default_value);
+
+    /**
+     * Reads an UTF-8 encoded string.
+     *
+     * @param fieldName name of the field.
+     * @return the value of the field.
+     * @throws hazelcast_serialization if the field does not exist in
+     * the schema or the type of the field does not match with the one defined
+     * in the schema.
+     */
+    boost::optional<std::string> read_string(const std::string& field_name);
+
+    /**
+     * Reads an UTF-8 encoded string or returns the default value.
+     *
+     * @param fieldName    name of the field.
+     * @param defaultValue default value to return if the field with the given
+     * name does not exist in the schema or the type of the field does not match
+     * with the one defined in the schema.
+     * @return the value or the default value of the field.
+     */
+    boost::optional<std::string> read_string(
+      const std::string& field_name,
+      const boost::optional<std::string>& default_value);
+
+private:
+    compact_reader(pimpl::compact_stream_serializer& compact_stream_serializer,
+                   object_data_input& object_data_input,
+                   const pimpl::schema& schema);
+
+    friend compact_reader pimpl::create_compact_reader(
+      pimpl::compact_stream_serializer& compact_stream_serializer,
+      object_data_input& object_data_input,
+      const pimpl::schema& schema);
+
+    bool is_field_exists(const std::string& fieldName,
+                         enum pimpl::field_kind kind);
```

Comment:
I don't know if it is needed but we can make this function const like the `get_field_descriptor`.

Reply:
Fixed here 7d6214369fe1cd0b922d38d78739ef4d97303582